packages/client-sdk-nodejs/src/internal/grpc/client-timeout-interceptor.ts

```diff
@@ -64,48 +65,79 @@ class RetryUntilTimeoutInterceptor {
 
   public createTimeoutInterceptor(): Interceptor {
     return (options, nextCall) => {
-      // Replace default timeout with the desired overall timeout
-      // on the first time through and set the first incremental timeout
-      if (this.overallDeadline === undefined) {
-        this.overallDeadline = createNewDeadline(this.overallRequestTimeoutMs);
-        options.deadline = createNewDeadline(
-          this.responseDataReceivedTimeoutMs
-        );
-        this.logger.debug(`new deadline set to ${options.deadline.valueOf()}`);
-        return new InterceptingCall(nextCall(options));
-      }
+      this.logger.debug(
+        `Previous RPC: ${this.previousRpc ?? 'none'} | Incoming RPC: ${
+          options.method_definition.path
+        }`
+      );
 
       // If the received deadline is equal to the overall deadline, we've
-      // maxed out the retries and should cancel the request.
+      // maxed out the retries on a particular request and should cancel retries.
       const receivedDeadline = options.deadline;
       if (
         receivedDeadline !== undefined &&
         receivedDeadline === this.overallDeadline
       ) {
         this.logger.debug(
-          'Unable to successfully retry request within overall timeout, canceling request'
+          `Unable to successfully retry request within overall timeout of ${this.overallRequestTimeoutMs}ms, canceling request`
         );
         // reset overall deadline for next request
         this.overallDeadline = undefined;
+        this.previousRpc = undefined;
         const call = new InterceptingCall(nextCall(options));
         call.cancelWithStatus(
           status.CANCELLED,
-          'Unable to successfully retry request within overall timeout'
+          `Unable to successfully retry request within overall timeout of ${this.overallRequestTimeoutMs}ms`
         );
         return call;
       }
 
+      // Reset overall and incremental deadlines in these cases:
+      // Case 1: overallDeadline is undefined (first request or previous was canceled)
+      // Case 2: different RPC path requested through same client
+      // Case 3: new request through same client after original retry deadline has passed
+      // (Note: Case 3 check must occur after the receivedDeadline == options.deadline check
+      // or else the request is always retried).
+      if (
+        this.overallDeadline === undefined ||
```

Comment:
oh gosh, I hadn't thought about the scope of this variable. If this interceptor is basically global to the client object, and we're trying to store this overallDeadline in a member variable, then we only have 1 of those variables for the whole client. That won't work, the client could be handling 100s of concurrent requests.

We'll need a way to have a separate copy of that variable for each request. I think you might be able to do that by defining that var using a `let` to create a local variable inside of the callback function, so that the deadline logic is basically a closure over the local variable.

Another option might be to store it in the metadata of the request, but we don't really want it to get sent up to the server. maybe there's some other place on the request object that we can attach metadata?

Failing that I think we would need to maintain a map where each request had its own key/value, but that will be tricky to make sure we clean it up properly so that we don't leak memory.

:(

very glad you caught this though. We should think about how to test this with two concurrent requests.

I can pull the code down again in a second and see if I can help figure out a way to store the variable as a local inside of the closure rather than a class instance variable.